ocaml/xapi/xapi_ha_vm_failover.ml

```diff
@@ -175,60 +197,460 @@ let order_f (_, vm_rec) =
+    =
+  match HostsSet.min_elt_opt hosts_set with
+  | Some (0, host_size, host) -> (
+    match (vm_size <= host_size, vm_can_boot_on_host vm host) with
+    | true, true ->
+        Some host
+    | _, _ ->
+        select_host_for_no_breach_plan vm vm_size vm_can_boot_on_host
+          (HostsSet.remove (0, host_size, host) hosts_set)
+  )
+  | _ ->
+      None
+
+(** Try to get a spread evenly plan for anti-affinity VMs (for each anti-affinity group, the number of 
+    running VMs from that group are spread evenly in all the rest hosts in the pool):
+    1. For all the anti-affinity VMs which sorted in an increasing order in terms of the VM's memory
+       size, do host selection as below step 2.
+    2. For each anti-affinity VM, select a host which can run it, and which has the least VMs in the
+       same anti-affinity group running on it, for the hosts with the same number of running VMs in
+       that group, pick the one with the least free memory. *)
+let rec compute_spread_evenly_plan ~__context pool_state vm_can_boot_on_host
+    anti_affinity_vms mapping =
+  debug "compute_spread_evenly_plan" ;
+  match anti_affinity_vms with
+  | [] ->
+      mapping
+  | (vm, vm_size, group) :: remaining_vms -> (
+      debug
+        "Spread evenly plan: try to plan for the anti-affinity VM (%s %s %s)."
+        (Ref.string_of vm)
+        (Db.VM.get_name_label ~__context ~self:vm)
+        (Db.VM_group.get_name_label ~__context ~self:group) ;
+      match
+        VMGroupMap.find group pool_state.grp_hosts_set
+        |> select_host_for_spread_evenly_plan vm vm_size vm_can_boot_on_host
+      with
+      | None ->
+          debug
+            "Spread evenly plan: no host can run this anti-affinity VM. Stop \
+             the planning as there won't be a valid plan for this VM." ;
+          raise (Api_errors.Server_error (Api_errors.no_hosts_available, []))
+      | Some h ->
+          debug
+            "Spread evenly plan: choose the host with the least free memory \
+             which can run the vm: (%s %s)."
+            (Ref.string_of h)
+            (Db.Host.get_name_label ~__context ~self:h) ;
+          compute_spread_evenly_plan ~__context
+            (pool_state |> update_spread_evenly_pool_state vm_size group h)
+            vm_can_boot_on_host remaining_vms ((vm, h) :: mapping)
+    )
+
+(** Try to get a no breach plan for anti-affinity VMs (for each anti-affinity group, there are at least
+    2 hosts having running VMs in the group): 
+    1. For all the anti-affinity VMs which sorted in an increasing order in terms of the VM's memory
+       size, do host selection as below step 2.
+    2. For each anti-affinity VM, try to select a host for it so that there are at least 2 hosts which
+       has running VMs in the same anti-affinity group. If there are already 2 hosts having running VMs
+       in that group, skip planning for the VM. *)
+let rec compute_no_breach_plan ~__context total_hosts pool_state
```

Comment:
How to make sure every group resides on at least 2 hosts eventually?

Reply:
No, we can't, we just try to.
The logic here is: for any anti-affinity VM, let's say an anti-affinity VM1, and it is from group A, if there are not 2 or more than 2 hosts which having running VMs from group A, we just try our best to plan it in some "no resident host"(a host which does not have any running VM from group A resident on it), but if we can't find such a "no resident host" for VM1, we will not stop `compute_no_breach_plan` and fail the no breach plan, as even if we fail to make a no breach plan for group A after `compute_no_breach_plan` is done for all the anti-affinity VMs, we might got a no breach plan for other groups.

And the same applies to `compute_spread_evenly_plan`, we can't make sure the VM's are spread evenly for each group finally.

Reply:
I understand for `spead_evenly` all will fail if any one group fails. While for `no_breach`, one group fails doesn't mean all fails.
Why is there is a different policy?

Reply:
> I understand for spead_evenly all will fail if any one group fails. While for no_breach, one group fails doesn't mean all fails.

Right, this is what I explained yesterday: for spread evenly plan, any anti-affinity VM should be planned on some host. While for no breach plan, it doesn't need to. As for "no breach planning", we only tried on the "no resident hosts".

- - - - - - - - -

> we can't make sure the VM's are spread evenly for each group finally.

Sorry for the mis-understanding, the meaning here is slightly different here: I mean the host selected for an anti-affinity VM might not be the "smallest" host in the `HostsSet`, as it might don't have enough free memory and can boot the VM. So although a host is selected for each anti-affinity VM, they might not spread "evenly" in the pool.